### PR TITLE
Contract call optimization - remove round trip

### PR DESCRIPTION
**Pull request by medvedev1088**
### What was wrong?

When calling a contract function on latest block, the lib unnecessarily calls getBlock('latest')

### How was it fixed?

Remove a round trip when block number is latest, earliest or pending when calling a contract function.

### CODE REVIEW

**Comment by pipermerriam** [comment]
Test failure appears unrelated (due to import sorting).

**Comment by voith** [comment]
@pipermerriam Test failures have been resolved by @carver  in https://github.com/ethereum/web3.py/pull/893. However, it is waiting to be merged.

File: web3/contract.py [hunk 1]
```diff
@@ -1397,12 +1397,15 @@ def call_contract_function(
...
-    if isinstance(block_identifier, int) and abs(block_identifier) <= last_block:
-        if block_identifier >= 0:
-            return block_identifier
+    if isinstance(block_identifier, int):
+        last_block = web3.eth.getBlock('latest').number
+        if abs(block_identifier) <= last_block:
+            if block_identifier >= 0:
+                return block_identifier
+            else:
+                return last_block + block_identifier + 1
```
Comment (carver) [inline web3/contract.py]: If the chain is 3 blocks long (so the latest number is 2), then -3 should be a valid identifier. One way to fix that edge case, and hopefully in a clean way to read, is:

```py
last_block = web3.eth.getBlock('latest').number

if block_identifier >= 0:
  block_num = block_identifier
else:
  block_num = last_block + block_identifier + 1

if 0 <= block_num <= last_block:
  return block_num
else:
  raise BlockNumberOutofRange
```

This is just long enough that it could go into its own method, like `parse_block_identifier_int()` and keep `parse_block_identifier` as a super simple dispatch method.

**Comment by medvedev1088** [comment]
@carver I fixed the edge case as you suggested, but unfortunately I will not have time to write tests for it.

BTW, what do you think about submitting a feature request to geth/parity to support negative block identifiers? In this case there will be no additional round trips even when block identifier is integer.

**Comment by carver** [comment]
> I fixed the edge case as you suggested, but unfortunately I will not have time to write tests for it.

Thanks, as soon as someone gets time to add the test, we'll get this merged.

> BTW, what do you think about submitting a feature request to geth/parity to support negative block identifiers? In this case there will be no additional round trips even when block identifier is integer.

Yeah, definitely worth a discussion. :+1: Feel free to @ mention me if you decide to open up issues and I'll +1 it

**Comment by dylanjw** [comment]
Didnt mean to add the breaking change label.  Hand must have slipped.

File: tests/core/contracts/test_contract_util_functions.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+from web3.contract import (
+    parse_block_identifier_int,
+)
+from web3.exceptions import (
+    BlockNumberOutofRange,
+)
+
+def test_parse_block_identifier_int(web3):
+    with pytest.raises(BlockNumberOutofRange):
+        parse_block_identifier_int(web3, -2)
```
[the remark this answers is not in the document]
Comment (carver) [reply tests/core/contracts/test_contract_util_functions.py]: No, as I read it, `-1 - w3.eth.getBlock('latest').number` should be block zero.

Say there are currently 3 blocks:

Block Number | Negative Index | Special
--- | --- | ---
0 | -3 | 
1 | -2 |
2 | -1 | latest

I think the test could be:
```py
last_num = w3.eth.getBlock('latest').number
assert w3.eth.getBlock(0) == w3.eth.getBlock(-1 - last_num)
```

**Comment by dylanjw** [comment]
@carver ok to merge?

**Comment by carver** [comment]
(except for the tests, of course ;))

File: tests/core/contracts/test_contract_util_functions.py [hunk 3]
```diff
@@ -1,3 +1,9 @@
+from web3.contract import (
+    parse_block_identifier_int,
+)
+
+
 def test_parse_block_identifier_int(web3):
     last_num = web3.eth.getBlock('latest').number
-    assert web3.eth.getBlock(0) == web3.eth.getBlock(-1 - last_num)
+    assert web3.eth.getBlock(0) == web3.eth.getBlock(
+        parse_block_identifier_int(web3, -1 - last_num))
```
[the remark this answers is not in the document]
Comment (dylanjw) [reply tests/core/contracts/test_contract_util_functions.py]: web3.eth.getBlock doesnt accept negative block identifiers.  parse_block_identifier_int will translate negative identifiers into the block number.  Maybe the assertion should just be:

```
assert 0 == parse_block_identifier_int(web3, -1 - last_num)
```

[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/core/contracts/test_contract_util_functions.py]: It's very strange that `parse_block_identifier` would normalize a negative number to zero.  That strikes me as unexpected behavior and likely to result in accidental errors.  It seems like if the number if negative then it should raise a `ValidationError` or something like that.

[the remark this answers is not in the document]
Comment (medvedev1088) [reply tests/core/contracts/test_contract_util_functions.py]: `parse_block_identifier` has logic to handle negative block numbers in a similar way Python does it for lists, e.g. `parse_block_identifier(-2)` will return the block before the last one. 
Currently this method would call getBlock('latest') for every invocation of a contract. What this pull request does is eliminates this extra call when block identifier (for the contract function call) is not numeric, e.g. 'latest'.
In my opinion, handling negative block identifiers for contract function calls should be removed for 2 reasons: (1) as @pipermerriam said it's unexpected behavior, (2) it's inconsistent with `web3.getBlock()` which doesn't handle negative block numbers. The problem with removing this logic is it will break backward compatibility.

[the remark this answers is not in the document]
Comment (carver) [reply tests/core/contracts/test_contract_util_functions.py]: Yeah, it seems to have been added in https://github.com/ethereum/web3.py/pull/663/commits/9f23db085e2405dd29c280f1b9ca5c6c65ee93f8

I don't see any rationale in the related PR or issue for why a negative index was particularly helpful or important. I'm happy to see it get dropped in v5. I also understand why this test did what it did now. So :+1: from me to merge this PR (which isn't making the situation worse) as is. 

Although it's probably worth adding a comment on the test, mentioning that it is converting the negative index from the latest block back to zero. Also saying that the test has to do the conversion explicitly because transaction calls allow this negative index, but `getBlock()` doesn't.
